internal/mock/cloudsql.go

```diff
@@ -77,6 +80,72 @@ func generateInstanceCerts() (*rsa.PrivateKey, *x509.Certificate, error) {
 		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
 		BasicConstraintsValid: true,
 	}
+	certBytes, err := x509.CreateCertificate(
+		rand.Reader, cert, cert, &key.PublicKey, key)
+	if err != nil {
+		return nil, nil, tls.Certificate{}, err
+	}
+
+	caPEM := &bytes.Buffer{}
+	pem.Encode(caPEM, &pem.Block{Type: "CERTIFICATE", Bytes: certBytes})
+
+	caKeyPEM := &bytes.Buffer{}
+	pem.Encode(caKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(key),
+	})
+
+	serverCert, err := tls.X509KeyPair(caPEM.Bytes(), caKeyPEM.Bytes())
+	if err != nil {
+		return nil, nil, tls.Certificate{}, err
+	}
 
-	return privKey, cert, nil
+	return key, cert, serverCert, nil
+}
+
+// ServerProxyConfig holds configuration values for the proxy.
+type ServerProxyConfig struct {
+	// Response is the value the proxy will write back to the client.
+	Response string
+	// Instance is the Fake Cloud SQL instance used for TLS configuration.
+	Instance FakeCSQLInstance
+	// InvalidCert determines if the server proxy should start up with a known
+	// bad configuration.
+	InvalidCert bool
+}
+
+// StartServerProxy starts a fake server proxy and listens on the provided port
+// on all interfaces, configured with TLS as specified by the FakeCSQLInstance.
+// Callers should invoke the returned function to clean up all resources.
+func StartServerProxy(config ServerProxyConfig) func() {
+	cert := config.Instance.tlsCert
+	if config.InvalidCert {
+		cert = tls.Certificate{}
+	}
```

Comment:
What's the point of this logic? 

Can we just update the instance to have the config instead of relying on a config here?

Reply:
As in pass the instance in as a parameter here? I think I like that. It does mean we'll have to export the fields on the instance.